pkg/service/id_local.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var (
+	mutex      lock.RWMutex
+	servicesID = map[uint32]*types.L3n4AddrID{}
+	services   = map[string]uint32{}
+	nextID     = common.FirstFreeServiceID
+	maxID      = common.MaxSetOfServiceID
+)
+
+func newServiceID(svc types.L3n4Addr, id uint32) *types.L3n4AddrID {
+	return &types.L3n4AddrID{
+		L3n4Addr: svc,
+		ID:       types.ServiceID(id),
+	}
+}
+
+func addServiceID(svc types.L3n4Addr, id uint32) *types.L3n4AddrID {
+	svcID := newServiceID(svc, id)
+	servicesID[id] = svcID
+	services[svc.String()] = id
+
+	return svcID
+}
+
+func acquireLocalID(svc types.L3n4Addr, desiredID uint32) (*types.L3n4AddrID, error) {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if svcID, ok := services[svc.String()]; ok {
```

Comment:
Be aware `svc.String()` might not be what you want. `.String()` will only return IP and Port, it will **not** return the service's protocol. Maybe "SHA256Sum()"?

Reply:
Good catch. String() should include the protocol

Reply:
fixed